titanic.py

```diff
@@ -86,19 +86,58 @@ def conditional_mutual_information(p, j, *conditional_indices):
     return (conditional_shannon_entropy(np.sum(p, axis=j), *marginal_conditional_indices)
             - conditional_shannon_entropy(p, j, *conditional_indices))
 
+
 def maximum_energy_delta(bqm):
     """Compute conservative bound on maximum change in energy when flipping a single variable"""
     return max(abs(bqm.get_linear(i))
                + sum(abs(bqm.get_quadratic(i,j))
                      for j in bqm.iter_neighbors(i))
                for i in bqm.iter_variables())
 
-def run_demo():
+
+def mutual_information_bqm(dataset, features):
+    """Build a BQM that maximizes MI between survival and a subset of features"""
+    variables = ((feature, -mutual_information(prob(dataset[['survived', feature]].values), 1))
+                 for feature in features)
+    interactions = ((f0, f1, -conditional_mutual_information(prob(dataset[['survived', f0, f1]].values), 1, 2))
+                    for f0, f1 in itertools.permutations(features, 2))
+    return dimod.BinaryQuadraticModel(variables, interactions, 0, dimod.BINARY)
+
+
+def add_combination_penalty(bqm, k, penalty):
+    """Create a new BQM with an additional penalty biased towards k-combinations"""
+    kbqm = dimod.generators.combinations(bqm.variables, k, strength=penalty)
+    kbqm.update(bqm)
+    return kbqm
+
+
+def mutual_information_feature_selection(dataset, features):
+    """Run the MIFS algorithm on a D-Wave solver"""
+    
+    # Set up a QPU sampler that embeds to a fully-connected graph of all the variables
+    sampler = DWaveCliqueSampler()
+
+    # For each number of features, k, penalize selection of fewer or more features
+    selected_features = np.zeros((len(features), len(features)))
+
+    bqm = mutual_information_bqm(dataset, features)
+
+    # This ensures that the soltion will satisfy the constraints.
+    penalty = maximum_energy_delta(bqm)
+
+    for k in range(1, len(features) + 1):
+        kbqm = add_combination_penalty(bqm, k, penalty)
+        sample = sampler.sample(kbqm,
+                                label='Example - MI Feature Selection',
+                                num_reads=10000).first.sample
+        for fi, f in enumerate(features):
+            selected_features[k-1, fi] = sample[f]
+    return selected_features
+
+
+def run_demo(dataset):
```

Comment:
Changing this function so that it accepts the `dataset` argument seems to give the impression that it is a generally applicable function that could operate on an arbitrary dataset, but that's not really the case.  In particular, the function uses a hard-coded definition of the column name for the target ("survived"), the plotting code also has some labeling that is specific to this target name, and there is a hard-coded value of `keep = 8` that seems somewhat specific to this example also (and actually it's probably not necessary to pre-select down to 8 features in the first place, but that is a separate issue unrelated to this PR).

I think the best option would be one of:
1. Go back to original behavior where `run_demo` does not accept any arguments
2. Add the necessary additional arguments to make `run_demo` general (in particular, passing the target name)
3. Move the body of `run_demo` under `if __name__ == "__main__"`

I lean towards 3 because in my view `run_demo` acts more like a script than a function that might be re-used.

Reply:
This change was to allow running and testing in an interactive session.
The code I extracted out depends on __file__ which only is only defined when running as a script.
I would prefer option 2 which enables that and generalises to arbitrary dataframes

Reply:
Ah, yes jupyter notebooks seem to have issues with `__file__`.  FYI ipython is an alternative that doesn't seem to have those issues.

Your update with option 2 looks fine to me, but don't forget about the comment at the top of `read_demo` that is out of place.

Reply:
Comments fixed